File: backend/app/schemas/recommendations.py
from __future__ import annotations

from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class CropRecommendationRequest(BaseModel):
    soil_n: float = Field(ge=0)
    soil_p: float = Field(ge=0)
    soil_k: float = Field(ge=0)
    soil_ph: float = Field(ge=0, le=14)
    temperature_c: float
    humidity_pct: float = Field(ge=0, le=100)
    rainfall_mm: float = Field(ge=0)
    location: str
    season: Optional[str] = None
    historical_yield: Optional[float] = None


class CropRecommendationItem(BaseModel):
    crop: str
    confidence: float
    explanation: str


class CropRecommendationResponse(BaseModel):
    recommendations: List[CropRecommendationItem]
    model_version: str
    created_at: datetime
    cached: bool = False
    recommendation_id: Optional[str] = None


class PriceForecastRequest(BaseModel):
    crop: str
    market: str
    currency: str = "INR"


class PriceForecastSeries(BaseModel):
    horizon_days: int
    dates: List[date]
    forecast: List[float]
    lower: List[float]
    upper: List[float]


class PriceHistoricalSeries(BaseModel):
    dates: List[date]
    prices: List[float]


class PriceForecastResponse(BaseModel):
    crop: str
    market: str
    currency: str
    series: List[PriceForecastSeries]
    historical: Optional[PriceHistoricalSeries] = None
    mape: float
    confidence_interval: dict
    model_version: str
    created_at: datetime
    cached: bool = False
    recommendation_id: Optional[str] = None


class WeatherDay(BaseModel):
    date: date
    rainfall_mm: float = Field(ge=0)
    temperature_c: float


class WaterOptimizationRequest(BaseModel):
    crop: str
    growth_stage: str
    soil_moisture_pct: float = Field(ge=0, le=100)
    water_source: str
    field_area_acres: float = Field(gt=0)
    forecast: List[WeatherDay] = Field(min_length=1)


class IrrigationScheduleItem(BaseModel):
    date: date
    irrigation_mm: float
    irrigation_liters: float
    reason: str


class WaterOptimizationResponse(BaseModel):
    crop: str
    schedule: List[IrrigationScheduleItem]
    total_volume_liters: float
    water_savings_percent: float
    notes: List[str]
    model_version: str
    created_at: datetime
    cached: bool = False
    recommendation_id: Optional[str] = None

File: frontend/src/services/recommendations.ts
import { cachedPost } from "./api";

export interface CropRecommendationRequest {
  soil_n: number;
  soil_p: number;
  soil_k: number;
  soil_ph: number;
  temperature_c: number;
  humidity_pct: number;
  rainfall_mm: number;
  location: string;
  season?: string;
  historical_yield?: number | null;
}

export interface CropRecommendationItem {
  crop: string;
  confidence: number;
  explanation: string;
}

export interface CropRecommendationResponse {
  recommendations: CropRecommendationItem[];
  model_version: string;
  created_at: string;
  cached: boolean;
  recommendation_id?: string;
}

export interface PriceForecastRequest {
  crop: string;
  market: string;
  currency: string;
}

export interface PriceForecastSeries {
  horizon_days: number;
  dates: string[];
  forecast: number[];
  lower: number[];
  upper: number[];
}

export interface PriceHistoricalSeries {
  dates: string[];
  prices: number[];
}

export interface PriceForecastResponse {
  crop: string;
  market: string;
  currency: string;
  series: PriceForecastSeries[];
  historical?: PriceHistoricalSeries | null;
  mape: number;
  confidence_interval: {
    level: number;
    description: string;
  };
  model_version: string;
  created_at: string;
  cached: boolean;
  recommendation_id?: string;
}

export interface WeatherDay {
  date: string;
  rainfall_mm: number;
  temperature_c: number;
}

export interface WaterOptimizationRequest {
  crop: string;
  growth_stage: string;
  soil_moisture_pct: number;
  water_source: string;
  field_area_acres: number;
  forecast: WeatherDay[];
}

export interface IrrigationScheduleItem {
  date: string;
  irrigation_mm: number;
  irrigation_liters: number;
  reason: string;
}

export interface WaterOptimizationResponse {
  crop: string;
  schedule: IrrigationScheduleItem[];
  total_volume_liters: number;
  water_savings_percent: number;
  notes: string[];
  model_version: string;
  created_at: string;
  cached: boolean;
  recommendation_id?: string;
}

export const fetchCropRecommendation = async (
  payload: CropRecommendationRequest
): Promise<CropRecommendationResponse> => {
  const key = `crop:${JSON.stringify(payload)}`;
  return cachedPost<CropRecommendationResponse, CropRecommendationRequest>(
    key,
    "/recommendations/crop",
    payload
  );
};

export const fetchPriceForecast = async (
  payload: PriceForecastRequest
): Promise<PriceForecastResponse> => {
  const key = `price:${JSON.stringify(payload)}`;
  return cachedPost<PriceForecastResponse, PriceForecastRequest>(
    key,
    "/recommendations/price-forecast",
    payload
  );
};

export const fetchWaterOptimization = async (
  payload: WaterOptimizationRequest
): Promise<WaterOptimizationResponse> => {
  const key = `water:${JSON.stringify(payload)}`;
  return cachedPost<WaterOptimizationResponse, WaterOptimizationRequest>(
    key,
    "/recommendations/water-optimization",
    payload
  );
};
